git/libgit2/transport/ssh.go

```diff
@@ -295,15 +295,53 @@ func (t *sshSmartSubtransport) Action(transportOptionsURL string, action git2go.
 	return t.currentStream, nil
 }
 
-func (t *sshSmartSubtransport) createConn(addr string, sshConfig *ssh.ClientConfig) error {
+func (t *sshSmartSubtransport) createConn(addr string, proxyOpts *git2go.ProxyOptions, sshConfig *ssh.ClientConfig) error {
 	ctx, cancel := context.WithTimeout(context.TODO(), sshConnectionTimeOut)
 	defer cancel()
 
 	t.logger.V(traceLevel).Info("dial connection")
-	conn, err := proxy.Dial(ctx, "tcp", addr)
+
+	var conn net.Conn
+	var err error
+
+	if proxyOpts != nil {
+		switch proxyOpts.Type {
+		case git2go.ProxyTypeSpecified:
+			proxyUrl, err := url.Parse(proxyOpts.Url)
+			if err != nil {
+				return err
+			}
+			dialer, err := proxy.FromURL(proxyUrl, proxy.Direct)
+			if err != nil {
+				return err
+			}
+
+			// Try to use a ContextDialer, but fall back to a Dialer if that goes south.
+			ctxDialer, ok := dialer.(proxy.ContextDialer)
+			if ok {
+				conn, err = ctxDialer.DialContext(ctx, "tcp", addr)
+			} else {
+				conn, err = dialer.Dial("tcp", addr)
+			}
+		case git2go.ProxyTypeAuto:
+			dialer := proxy.FromEnvironment()
+			ctxDialer, ok := dialer.(proxy.ContextDialer)
+
+			if ok {
+				conn, err = ctxDialer.DialContext(ctx, "tcp", addr)
+			} else {
+				conn, err = dialer.Dial("tcp", addr)
+			}
+		default:
+			conn, err = net.Dial("tcp", addr)
+		}
+	} else {
+		conn, err = net.Dial("tcp", addr)
+	}
```

Comment:
`net.Dial` does not use context, so we would need to ensure Timeouts are enforced in a different manner.
Otherwise we will be breaking the behaviour established on: https://github.com/fluxcd/source-controller/pull/740

Reply:
I think we should get rid of this altogether. If `proxyOpts.Type == git2go.ProxyTypeSpecified` then we do `dialer, err := proxy.FromURL()` else `dialer, err := proxy.FromEnviornment()`, since that'll just do a regular `net.Dial()` when no env vars are present.

Reply:
or we could just use `net.DialTimeout()`, if we want to honor all possible values of `proxyOpts` (which is not unreasonable since this is a public libgit2 package)

Reply:
From a context enforcement that sounds reasonable. 

Overall users would be a bit more in control as to when they want to auto-detect proxy settings enabled, as opposed to the current "always on".

How would this be exposed to the end user? Would they have a `ProxyTypeAuto` set by default, and then have the ability to pick a per object setting? On that point, can we do something similar for `go-git`?